### PR TITLE
Add filename to result

### PR DESCRIPTION
**Pull request by cuff-links**
This PR adds the data for the filenames of the images that were created to the returned results. This is helpful when debugging test failures so you can output whatever exact image failed. If you have 5 different viewports, it can be helpful for displaying an image or pushing the image to some kind of cloud solution.

### CODE REVIEW

**Comment by cuff-links** [comment]
@zinserjan Added the ability to return file data from the createReport call. Modified tests to account for this. All is green. Let me know what you think.

File: package.json
Comment (emilyrohrbough) [inline package.json]: Why was this move to a dependency?

Comment (cuff-links) [reply package.json]: Hmm. I think this happened when I used yarn to add it. I will switch.

**Comment by cuff-links** [comment]
Thanks for catching that, @emilyrohrbough.

File: src/methods/BaseCompare.js
Comment (zinserjan) [inline src/methods/BaseCompare.js]: Chaning the order of arguments is a breaking change for all guys who uses custom compare methods. I would like to avoid this. Moreover this is to inflexibel and only useful for file comparison. I suggest to replace filePaths with an object that wil be assigned via Object.assign to the returned object. Then custom or other compare methods can enhance the report with additional information.

File: src/methods/LocalCompare.js
Comment (zinserjan) [inline src/methods/LocalCompare.js]: use const and assign value directly

File: src/methods/LocalCompare.js
Comment (zinserjan) [inline src/methods/LocalCompare.js]: - add diff property with value of null to make sure the property is always there
- replace key `actual` with `screenshot` to be consistent with the `referenceName  `, `screenshotName` and `diffName` option.

**Review by zinserjan — CHANGES_REQUESTED** [review]
Thanks for your work. I like the idea, but I think this needs to be done in a different way to make this more generic. See my comments.

File: src/methods/BaseCompare.js
Comment (cuff-links) [reply src/methods/BaseCompare.js]: @zinserjan Would you rather I made it the last argument, then? That way, if I am remembering my JS correctly, the value would just be null for those who aren't using the functionality?

File: src/methods/LocalCompare.js
Comment (cuff-links) [reply src/methods/LocalCompare.js]: Done.

**Comment by cuff-links** [comment]
@zinserjan I looked at the travis build and the failure seems a little funky. Passed on node 6 failed on 8. And it didn't have anything to do with test logic failing. Unfortunately, I can't rerun the tests to see if this was just a fluke failure.

**Comment by cuff-links** [comment]
@zinserjan @emilyrohrbough Can someone rerun this?

**Comment by cuff-links** [comment]
Is this PR still missing something? I believe all points in the review were addressed. @zinserjan @emilyrohrbough

**Comment by emilyrohrbough** [comment]
@silne30 I don't have privs to rerun tests.

**Comment by cuff-links** [comment]
Gotcha. Thanks.

**Comment by zinserjan** [comment]
@silne30 just restarted the test.
 I'll review this PR later ;)

File: src/methods/Spectre.js
Comment (zinserjan) [inline src/methods/Spectre.js]: Remove filesnames from spectre report. This method doesn't use any files.

File: src/methods/BaseCompare.js
Comment (zinserjan) [inline src/methods/BaseCompare.js]: > Moreover this is to inflexibel and only useful for file comparison. I suggest to replace filePaths with an object that wil be assigned via Object.assign to the returned object. Then custom or other compare methods can enhance the report with additional information.

This method should be more generic like I already said. What I mean is the following:

```js
  createResultReport(misMatchPercentage, isWithinMisMatchTolerance, isSameDimensions, options = {}) {
    return {
      ...options,
      misMatchPercentage,
      isWithinMisMatchTolerance,
      isSameDimensions,
      isExactSameImage: misMatchPercentage === 0,
    };
  }
````
This allows you to pass everything you want in and just merges this into the report object.

In your case you would call it the following for LocalCompare & SaveScreenshot. All other methods shouldn't be touched.

```js
this.createResultReport(misMatchPercentage, true, isSameDimensions, {
  filenames: {
    reference: referencePath,
    screenshot: screenshotPath,
    diff: null,
  },
});
```

**Review by zinserjan — CHANGES_REQUESTED** [review]

File: src/methods/BaseCompare.js
Comment (cuff-links) [reply src/methods/BaseCompare.js]: OK. Thanks for that. Had to look into the spread operator in ES6 because I thought ...options was a typo. XD

**Comment by cuff-links** [comment]
@zinserjan I think this is more what you were looking for. ^_^

**Comment by cuff-links** [comment]
@zinserjan Is there anything else this needs?

**Comment by cuff-links** [comment]
@emilyrohrbough @zinserjan I know you all are busy. Just making sure that there is nothing else that I can do on my end for this PR.

File: src/methods/LocalCompare.js
Comment (emilyrohrbough) [inline src/methods/LocalCompare.js]: nit: extra line

File: src/methods/LocalCompare.js
Comment (emilyrohrbough) [inline src/methods/LocalCompare.js]: nit: extra line

File: src/methods/LocalCompare.js
Comment (emilyrohrbough) [inline src/methods/LocalCompare.js]: nit: the var name options is misleading here as these are report results not config options.

**Comment by emilyrohrbough** [comment]
This looks good to me. Can you explain the reason for the phantom dependency change?

**Comment by cuff-links** [comment]
@emilyrohrbough 
`phantomjs` was [deprecated](https://github.com/Medium/phantomjs/issues/447) in favor of `phantomjs-prebuilt`.

File: src/methods/BaseCompare.js
Comment (yuderekyu) [inline src/methods/BaseCompare.js]: Nit: to be safe, wouldn't we want to spread `options` first? In the future, we could potentially override the other object properties if we add duplicate properties to the passed in `fileNames` object.

Comment (cuff-links) [reply src/methods/BaseCompare.js]: @yuderekyu Sorry. I am a bit new to ES6 and all of it's goodness. Can you explain to me what you mean by "spread options first"? First where?

Comment (yuderekyu) [reply src/methods/BaseCompare.js]:
```javascript
return {
  ...options, //spread this first
  misMatchPercentage,
  isWithinMisMatchTolerance,
  isSameDimensions,
  isExactSameImage: misMatchPercentage === 0,
};
```
Just to be safe. If `options` somehow contains any other keys (`misMatchPercentage`, `isWithinMismatchTolerance`, `isSameDimensions`, `isExactSameImage`), then spreading `options` last would override those values.

Comment (cuff-links) [reply src/methods/BaseCompare.js]: Gotcha.

**Comment by cuff-links** [comment]
@emilyrohrbough @yuderekyu @zinserjan We good now?

**Comment by cuff-links** [comment]
Ummmm.....can this be merged now?

**Comment by cuff-links** [comment]
??

**Comment by cuff-links** [comment]
@zinserjan I have many thumbs up here. Can this change be merged yet

**Comment by mgibas** [comment]
Thumbocracy 🤣
For real - maintainers either merge or reject - please please please :)

**Comment by cuff-links** [comment]
@zinserjan @emilyrohrbough @yuderekyu Is there anything I may have missed?

**Comment by emilyrohrbough** [comment]
@silne30 I do not have the permissions to merge this PR. I use the `wdio-visual-regression-service` within a project which is why I have been active on this PR.

**Comment by cuff-links** [comment]
@zinserjan Hitting the 90 day mark since last fix was made. Anything else this needs?

**Comment by cuff-links** [comment]
Hope everything is okay. Please approve if this works. Please give me a feedback if I need to change something.

**Comment by cuff-links** [comment]
@zinserjan Anyone?

**Comment by cuff-links** [comment]
Not sure if this project has nodded off. We are looking at almost 9 months since feedback from last commit. I am not even using this tool anymore but I am hoping the PR get's pulled for anyone else who may want the functionality. @zinserjan Are you still looking into this?

**Comment by cuff-links** [comment]
Gonna go ahead and close this. If we need it, we can reopen it. We are going on a year without any update since the last code change.

**Pull request closed without merging**
